k8s/policies.go

```diff
@@ -0,0 +1,238 @@
+		}
+
+		p.PodSecurityContext.DeepCopyInto(deployment.Spec.Template.Spec.SecurityContext)
+	}
+
+	return deployment
+}
+
+// Remove is the inverse of Apply, removing the mutations that the Policy would have applied
+func (p Policy) Remove(deployment *appsv1.Deployment) *appsv1.Deployment {
+
+	for _, policyToleration := range p.Tolerations {
+		// filter the existing tolerations and then update the deployment
+		// filter without allocation implementation from
+		// https://github.com/golang/go/wiki/SliceTricks#filtering-without-allocating
+		newTolerations := deployment.Spec.Template.Spec.Tolerations[:0]
+		for _, toleration := range deployment.Spec.Template.Spec.Tolerations {
+			if !reflect.DeepEqual(policyToleration, toleration) {
+				newTolerations = append(newTolerations, toleration)
+			}
+		}
+		deployment.Spec.Template.Spec.Tolerations = newTolerations
+	}
+
+	if p.RuntimeClassName != nil {
+		if equalStrings(deployment.Spec.Template.Spec.RuntimeClassName, p.RuntimeClassName) {
+			deployment.Spec.Template.Spec.RuntimeClassName = nil
+		}
+	}
+
+	if p.Affinity != nil && reflect.DeepEqual(p.Affinity, deployment.Spec.Template.Spec.Affinity) {
+		deployment.Spec.Template.Spec.Affinity = nil
+	}
+
+	if p.PodSecurityContext != nil {
+		sc := deployment.Spec.Template.Spec.SecurityContext
+
+		if reflect.DeepEqual(p.PodSecurityContext.SELinuxOptions, sc.SELinuxOptions) {
+			deployment.Spec.Template.Spec.SecurityContext.SELinuxOptions = nil
+		}
+		if reflect.DeepEqual(p.PodSecurityContext.SELinuxOptions, sc.WindowsOptions) {
+			deployment.Spec.Template.Spec.SecurityContext.WindowsOptions = nil
+		}
+		if p.PodSecurityContext.RunAsUser != nil {
+			deployment.Spec.Template.Spec.SecurityContext.RunAsUser = nil
+		}
+		if p.PodSecurityContext.RunAsGroup != nil {
+			deployment.Spec.Template.Spec.SecurityContext.RunAsGroup = nil
+		}
+		if p.PodSecurityContext.RunAsNonRoot != nil {
+			deployment.Spec.Template.Spec.SecurityContext.RunAsNonRoot = nil
+		}
+		if p.PodSecurityContext.SupplementalGroups != nil {
+			deployment.Spec.Template.Spec.SecurityContext.SupplementalGroups = nil
+		}
+		if p.PodSecurityContext.FSGroup != nil {
+			deployment.Spec.Template.Spec.SecurityContext.FSGroup = nil
+		}
+		if p.PodSecurityContext.Sysctls != nil {
+			deployment.Spec.Template.Spec.SecurityContext.Sysctls = nil
+		}
+	}
+
+	return deployment
+}
+
+func equalStrings(a, b *string) bool {
+	if a == nil && b == nil {
+		return true
+	}
+
+	if a != nil && b == nil {
+		return false
+	}
+
+	if a == nil && b != nil {
+		return false
+	}
+
+	// now we know both values are non-nil
+	return *a == *b
+}
+
+type policyClient struct {
+	kube typedCorev1.ConfigMapsGetter
+}
+
+func NewConfigMapPolicyClient(kube kubernetes.Interface) PolicyClient {
+	return &policyClient{kube: kube.CoreV1()}
+}
+
+// Get returns the named policies, if found, from the namespace
+func (c policyClient) Get(namespace string, names ...string) ([]Policy, error) {
```

Comment:
Is this code now run for every deployment and update?

If so, then it will likely cause the controller to be rate-limited when people do load tests or large deployments of functions within a short time window.

We use an informer/lister pattern for deployments and it would be beneficial here too.